### PR TITLE
refactor substance page for compound saving in substance form

### LINKED ISSUE #122: VueJS Store usage for substances and compounds
**As a** *developer*
**I want** *usage of the store and usage of component variables have a clear pattern and match with business logic*
**So that** *the code is clean and easy to understand and bugs are less likely*

**Acceptance Criteria:**
A plan for refactoring and use of the store is outlined and documented?
The code is refactored to match the plan
The changes are not apparent to the user (i.e. no bugs introduced)

**Test(s):**

### LINKED ISSUE #135: Substance save bug (dtxcid relationship not saving)
**Describe the bug**

**To Reproduce**
1. Start this a substance record
image

2. Alter the structure to a different compound (note it cannot be a compound that is linked to a different substance or the submit would fail)
image

3. See Save Substance is disabled (bug 1)
4. Alter another substance field like the name
5. Save the substance
6. See the structure associated with the substance reverts
image

**Expected behavior**
As a record compound id changes on the substance, this change should enable "Save Substance" (as long as the DTXCID in the record compound id is different than currently stored for that substance by the API).  When I press save, the associatedCompound should be updated if it has been changed.

**Screenshots**

**Desktop (please complete the following information):**

**Additional context**
The associated ticket #97 did not clarify that the record compound id should be considered the id associatedCompound.  Bad product owner for not making this clear, but this linkage must be updated as part of the save action.

### LINKED ISSUE #147: The substance being viewed should be reflected in URL
**As a** *user*
**I want** *the substance being viewed to be reflected in the URL *
**So that** *if I refresh that page, I will still be on the same record*

**Acceptance Criteria:**
After selecting a substance from the sidebar on the substance page or when arriving on the substance page via a search in the navbar searchbox, the URL reflects that and a browser refresh brings you to the same content.

**Test(s):**

### PR DESCRIPTION
**Pull request by debboutr**
closes #122
closes #135
closes #147

- [x]  use URL routing so that a page refresh will load the substance if one is in the form
`http://localhost:8080/substance/DTXSID202000002`
`http://localhost:8080/substance?search=hydrogen%20peroxide`
- [x]  add the compound to the substance form for saving
- [x] substance link on the navbar clears out the page data
- [x] PATCH update to QST can be done with Marvin
- [x]  "save substance" button enabled when appropriate
- [x] compound saves to the substance on PATCH or POST

There is a whole lot going on in this branch, the PR in django [#312](https://github.com/Chemical-Curation/chemcurator_django/pull/312) is needed for the tests to pass here.

### CODE REVIEW

File: src/components/ChemicalEditors.vue [hunk 1]
```diff
@@ -103,28 +107,25 @@ export default {
...
+        // this block empties out the ketcher window when substance is
+        // hit in the NavBar
+        this.$refs["ketcher"].loadMolfile("");
+        this.changed = false;
       } else if (this.initialCompound?.type === "definedCompound") {
-        this.definedCompound = this.initialCompound;
-        this.type = "definedCompound";
-        // Attempt to load the new molfile
-        this.$refs["ketcher"].loadMolfile(this.initialMolfile);
+        console.log("watcher", this.definedCompound.attributes.molfileV3000);
```
Comment (michael-barbour) [inline src/components/ChemicalEditors.vue]: log

**edit: Added on a draft PR. Resolving.

File: src/components/KetcherWindow.vue [hunk 2]
```diff
@@ -61,15 +58,22 @@ export default {
       );
     },
     clearMolfile: function() {
+      console.log("clearing...");
```
Comment (michael-barbour) [inline src/components/KetcherWindow.vue]: log

**edit: Added on a draft PR. Resolving.

File: src/components/KetcherWindow.vue [hunk 3]
```diff
@@ -36,15 +35,13 @@ export default {
...
-      loadedMolfile: ""
+      molfile: ""
     };
   },
   methods: {
     loadMolfile: function(molfile) {
       if (molfile) {
-        this.loadedMolfile = "";
-
+        console.log("loading...", molfile);
```
Comment (michael-barbour) [inline src/components/KetcherWindow.vue]: log

Comment (michael-barbour) [reply src/components/KetcherWindow.vue]: Added on a draft PR.  Resolving.

**Comment by hausman-gdit** [comment]
It looks like clearing the marvin window was removed.

"Sample Substance 2" -> search "Solo Substance" , there is still an illdefined compound

149-ill-defined

**Comment by michael-barbour** [comment]
Looking at this solution, I'm not sure it achieves the ask for "A plan for refactoring and use of the store is outlined and documented".  I was of the impression this ticket was supposed to facilitate team discussion and hopefully have us reach a more unified goal.

There was the offer to start a discussion last Thursday (01-07) that never actually got started, but I think that's a good idea and necessary for this ticket.

Some of my specific concerns are related to how code will have to be written using VueX as a data store and how non-reusable components become when we tie them to a specific piece of data within a VueX store.

File: src/components/substance/SubstanceForm.vue [hunk 4]
```diff
@@ -216,6 +258,8 @@ export default {
       let { id } = this.form;
       let payload = this.buildPayload();
       if (id) {
+        console.log("patch", this.$store.state.compound.type);
+        console.log("patch", payload);
```
Comment (michael-barbour) [inline src/components/substance/SubstanceForm.vue]: Log
```suggestion
```

File: src/components/substance/SubstanceForm.vue [hunk 5]
```diff
@@ -226,6 +270,7 @@ export default {
           .then(response => this.handleSuccess(response))
           .catch(err => this.handleFail(err));
       } else {
+        console.log("post", payload);
```
Comment (michael-barbour) [inline src/components/substance/SubstanceForm.vue]: Log
```suggestion
```

File: src/components/substance/SubstanceForm.vue [hunk 6]
```diff
@@ -234,10 +279,11 @@ export default {
       }
     },
     handleSuccess(response) {
+      console.log("response", response);
```
Comment (michael-barbour) [inline src/components/substance/SubstanceForm.vue]: Log
```suggestion
```

File: src/components/substance/SubstanceForm.vue [hunk 7]
```diff
@@ -249,6 +295,7 @@ export default {
     },
     handleFail(err) {
       // `sid` is included here to prevent it's input state from going true
+      console.log("errors", err.response);
```
Comment (michael-barbour) [inline src/components/substance/SubstanceForm.vue]: Log
```suggestion
```

File: src/components/substance/SubstanceForm.vue [hunk 8]
```diff
@@ -258,6 +305,13 @@ export default {
           .shift();
         if (attr == "nonFieldErrors") {
           nonField.push(error.detail);
+        } else if (attr == "associatedCompound") {
+          console.log(error.detail);
```
Comment (michael-barbour) [inline src/components/substance/SubstanceForm.vue]: Log
```suggestion
```

File: src/components/substance/SubstanceForm.vue [hunk 9]
```diff
@@ -271,12 +325,8 @@ export default {
           this.$set(this.validationState[field], "state", true);
         });
       // handle nonField errors
+      console.log("nonfield", nonField);
```
Comment (michael-barbour) [inline src/components/substance/SubstanceForm.vue]: Log
```suggestion
```

**Review by michael-barbour — CHANGES_REQUESTED** [review]
Console logs that were left in.  We use `npm build` and it has a rule that we cannot build with `console.logs`

**edit: Removed in `52f390c `

**Review by michael-barbour — APPROVED** [review]
Tests are passing with the seed data.

This resolves 3 tickets but it looks like all of their requirements are met here too.

**Pull request merged**